File: 3SumClosest.c
#include<stdio.h>
#include<string.h>
#include<stdlib.h>
#include<stdbool.h>
#include<math.h>
void quicksort(int l, int r, int *obj)
{
	if (l >= r)	return;
	int i = l, j = r, x = obj[r];
	while (i < j)
	{
		while (i < j && obj[i] <= x)	++ i;
		obj[j] = obj[i];
		while (i < j && obj[j] > x)		-- j;
		obj[i] = obj[j];
	}
	obj[i] = x;
	quicksort(l, i - 1, obj);
	quicksort(i + 1, r, obj);
}
int threeSumClosest(int *nums, int numsSize, int target)
{
	quicksort(0, numsSize - 1, nums);
	int abs(int x) {return x > 0 ? x : -x;}
	int out = nums[0] + nums[1] + nums[2];
	for (int i = 0; i < numsSize - 2; ++ i)
	{
		int j = i + 1, k = numsSize - 1;
		while (j < k)
		{
			int temp = nums[i] + nums[j] + nums[k];
			if (abs(target - temp) < abs(target - out))	out = temp;
			if (temp == target)	return target;
			if (temp < target)	++ j;
			if (temp > target)	-- k;
		}
	}
	return out;
}
int main()
{
	FILE *fin = fopen("oo.xx", "r");
	FILE *fout = fopen("xx.oo", "w");
	int m, n, a[500];
	fscanf(fin, "%d%d", &n, &m);
	for (int i = 0; i < n; ++ i )
		fscanf(fin, "%d", a + i);
	int out = threeSumClosest(a, n, m);
	printf("%d\n", out);
	fprintf(fout, "%d\n", out); 
	return 0;
}

File: 3SumClosest.py
from typing import List
class Solution:
    def quicksort(l, r, obj):
        if (l >= r):
            return
        i, j, x = l, r, obj[r]
        while i < j:
            while i < j and obj[i] <= x:
                i += 1
            obj[j] = obj[i]
            while i < j and obj[j] > x:
                j -= 1
            obj[i] = obj[j]
        obj[i] = x
        Solution.quicksort(l, i - 1, obj)
        Solution.quicksort(i + 1, r, obj)
    def threeSumClosest(self, nums: List[int], target: int) -> int:
        last = len(nums) - 1
        Solution.quicksort(0, last, nums)
        out = nums[0] + nums[1] + nums[2]
        for i, x in enumerate(nums):
            j, k = i + 1, last
            if j >= k:
                return out
            while j < k:
                temp = nums[i] + nums[j] + nums[k]
                if abs(out - target) > abs(temp - target):
                    out = temp
                if temp < target:
                    j += 1
                else:
                    k -= 1
        return out

fin = open("oo.xx", "r")
fout = open("xx.oo", "w")
target = int(fin.readline()[:-1].split(' ')[1])
nums = [int(x) for x in fin.readline()[:-1].split(' ')]
out = Solution().threeSumClosest(nums, target)
print(out)
fout.write(str(out))

